[optparse] Change Uses of List[] to Sequence[] in argument positions.

gvanrossum opened the pull request:
One of these (parse_args()) broke in real-world code, because the
expected type was List[Union[str, bytes]] but the actual type was
List[str] (because List is invariant).  That is ridiculous, so I
changed the accepted type to Sequence[_Text] which is covariant.  Then
I found a few other methods/functions that probably should also be
changed to Sequence, but I'm less certain of those.  I'm not at all
sure about the instance/class attributes, so I left those alone
(though I suspect those might also have to switch).

@JelleZijlstra and @ethans Thoughts? Am I going too far or not far enough?

JelleZijlstra commented:
This change looks reasonable, but I don't have time right now to look in detail at how optparse handles these arguments (I didn't get a chance to review Ethan's revised PR either).

gvanrossum commented:
@ethanhs If you have a moment I'd appreciate it.

ethanhs commented:
Let me take a look. Unless I am missing something, there is never an expected type of `List[Union[str, bytes]]`, so I think you meant something else? But I understand the issue.

My guess is that the member variables will have to be modified, as assignment to `List[]` from `Sequence[]` will cause problems. Back to reading through optparse.py I suppose! 😃

JelleZijlstra commented:
The `List[Union[str, bytes]]` is from expanding the `_Text` alias.

ethanhs commented:
Ah, it has been a while since I have used Python 2, I guess the confusion comes from this issue with mypy: https://github.com/python/mypy/issues/2968

gvanrossum commented:
So can I commit this or do you need to check something more carefully first?

ethans commented:
I know that I wasn't the Ethan originally intended for this pull request, but it seems to me like the correct fix for the issue.
And I'll take the opportunity to thank you for the fantastic job you're doing with Python. If there is a way that I can contribute - I'd love to.

Thank you,
Ethan

ethanhs commented on stdlib/2and3/optparse.pyi:
```diff
@@ -136,7 +136,7 @@ class OptionContainer:
     def _share_option_mappings(self, parser: OptionParser) -> None: ...
     def add_option(self, *args, **kwargs) -> Any: ...
-    def add_options(self, option_list: List[Option]) -> None: ...
+    def add_options(self, option_list: Sequence[Option]) -> None: ...
```
This is just iterated over in a for loop. So `Iterable[Option]` may be more correct.

ethanhs commented on stdlib/2and3/optparse.pyi:
```diff
@@ -171,7 +171,7 @@ class OptionParser(OptionContainer):
     version = ...  # type: _Text
     def __init__(self, usage: Optional[_Text] = ...,
-                       option_list: List[Option] = ...,
+                       option_list: Sequence[Option] = ...,
```
Also possibly better as `Iterable[Option]`, it is just passed onto the same method.

ethanhs commented on stdlib/2and3/optparse.pyi:
```diff
@@ -187,8 +187,8 @@ class OptionParser(OptionContainer):
     def _match_long_opt(self, opt: _Text) -> _Text: ...
-    def _populate_option_list(self, option_list: List[Option], add_help: bool = ...) -> None: ...
-    def _process_args(self, largs: List[_Text], rargs: List, values: Values) -> None: ...
+    def _populate_option_list(self, option_list: Sequence[Option], add_help: bool = ...) -> None: ...
```
This is called with `option_list` from `__init__` as an argument, so `Iterable[Option]` if you are changing the others.

ethanhs commented on stdlib/2and3/optparse.pyi:
```diff
@@ -187,8 +187,8 @@ class OptionParser(OptionContainer):
-    def _populate_option_list(self, option_list: List[Option], add_help: bool = ...) -> None: ...
-    def _process_args(self, largs: List[_Text], rargs: List, values: Values) -> None: ...
+    def _populate_option_list(self, option_list: Sequence[Option], add_help: bool = ...) -> None: ...
+    def _process_args(self, largs: Sequence[_Text], rargs: List, values: Values) -> None: ...
```
This is a bit more complex. In the code it calls `largs.append(...)`, so I believe sequence is too generic, it is looking for a list.

gvanrossum replied:
Since it's just an internal method I'm changing it back to plain `List` (implying `List[Any]`) -- that's what's used for `rargs` too, and they are pretty similar -- both must be lists (or MutableMappings?) of "strings" (per documentation).

ethanhs commented:
@gvanrossum I am almost certain it would be better to do `TextList = Union[List[str], List[unicode]]` for Python2, and `TextList = List[str]` in Python 3, then change `List[_Text]` to `TextList` everywhere. What do you think?

gvanrossum commented:
> I am almost certain it would be better to do `TextList = Union[List[str], List[unicode]]` for Python2, and `TextList = List[str]` in Python 3, then change `List[_Text]` to `TextList` everywhere. What do you think?

I don't like that, since that union would disallow lists of mixed (bytes/unicode) content, which AFAIK the code is fine with that.

gvanrossum commented:
(Sorry I accidentally pushed a bad commit, you might have received an email about that, I've rebased, removed the bad commit and re-pushed.) I think this is now fine -- @ethanhs what do you think?

ethanhs commented:
Fair point on mixed lists. Changes LGTM. 👍
